src/main/java/io/amtech/projectflow/repository/impl/ProjectCommentCustomRepositoryImpl.java

```diff
@@ -0,0 +1,61 @@
+package io.amtech.projectflow.repository.impl;
+
+import io.amtech.projectflow.app.general.PagedData;
+import io.amtech.projectflow.app.general.SearchCriteria;
+import io.amtech.projectflow.domain.project.Project;
+import io.amtech.projectflow.domain.project.ProjectComment;
+import io.amtech.projectflow.domain.project.ProjectComment_;
+import io.amtech.projectflow.domain.project.Project_;
+import io.amtech.projectflow.repository.ProjectCommentCustomRepository;
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Join;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.amtech.projectflow.util.DateUtil.secondsToInstant;
+import static io.amtech.projectflow.util.OrderUtil.parseOrder;
+
+@Repository
+public class ProjectCommentCustomRepositoryImpl implements ProjectCommentCustomRepository {
+
+    @PersistenceContext
+    private EntityManager em;
+
+    @Override
+    public PagedData<ProjectComment> search(long projectId, SearchCriteria criteria) {
+        var builder = em.getCriteriaBuilder();
+        CriteriaQuery<ProjectComment> query = builder.createQuery(ProjectComment.class);
+        Root<Project> root = query.from(Project.class);
+        Join<Project, ProjectComment> join = root.join(Project_.projectComments);
+        List<Predicate> predicates = new ArrayList<>() {{
```

Comment:
такую конструкцию желательно не использовать в продуктовом коде, хотя для написания тестов довольно удобно
Причины
```
Double Brace Initialization should not be used
 
Bug
 
Minor
java:S3599
 
Because Double Brace Initialization (DBI) creates an anonymous class with a reference to the instance of the owning object, its use can lead to memory leaks if the anonymous inner class is returned and held by other objects. Even when there's no leak, DBI is so obscure that it's bound to confuse most maintainers.
For collections, use Arrays.asList instead, or explicitly add each item directly to the collection.
Noncompliant Code Example
  Map source = new HashMap(){{ // Noncompliant      put("firstName", "John");      put("lastName", "Smith");  }};  
Compliant Solution
  Map source = new HashMap();  // ...  source.put("firstName", "John");  source.put("lastName", "Smith");  // ...  
```

Reply:
как вариант можно заменить конструкцией
```
Arrays.asList(builder.equal(root.get(Project_.ID), projectId));
```